pyccel/codegen/printing/ccode.py

```diff
@@ -323,7 +323,9 @@ def arrayFill(self, expr):
             sympy_shapes = [pyccel_to_sympy(s, symbol_map, used_names_tmp) for s in lhs.alloc_shape]
 
             length = functools.reduce(operator.mul, sympy_shapes)
-            length_code = '*'.join(self._print(i) for i in lhs.alloc_shape)
+            printable_length = functools.reduce(PyccelMul, lhs.alloc_shape)
+
+            length_code = self._print(printable_length)
```

Comment:
Do we still need the `sympy_shapes` variable?

Reply:
Yes `length` and `length_code` are 2 different objects. `length` is a sympy object containing the length so we can tell if the length is known at compile time or not. `length_code` is the code that will be printed for the array size. If the `length` is constant then the code is generated differently. E.g:
1.
```
@stack_array('a')
def f():
    a = np.empty(3)
```
generates:
```
void f(void)
{
    t_ndarray a;
    a = (t_ndarray){
        .nd_double=(double[3]){},
        .shape=(int64_t[]){3},
        .strides=(int64_t[1]){0},
        .nd=1,
        .type=nd_double,
        .is_view=false
    };
    stack_array_init(&a);
}
```
which is compact and requires no additional variables.
2. For the following example:
```
@stack_array('a')
def f(d : int):
    a = np.empty(d)
```
the same cannot be done as a literal cannot have unknown size. We must therefore do:
```
void f(int64_t d)
{
    t_ndarray a;
    double a_data_0001[d];
    a = (t_ndarray){
        .nd_double=a_data_0001,
        .shape=(int64_t[]){d},
        .strides=(int64_t[1]){0},
        .nd=1,
        .type=nd_double,
        .is_view=false
    };
    stack_array_init(&a);
}
```

This was the subject of PR #841

Reply:
This may be more complicated than it needs to be though? Perhaps it is sufficient to do an atomic search for an instance of a `Variable`?

Reply:
I was thinking that maybe, once we simplify `lhs.alloc_shape`, it should reduce to a `Literal` object if it is a constant. Wishful thinking on my side?

Reply:
I think so. While something like `2+2` should reduce to a `Literal` something more complicated like: `(1+d)*(1-d)+d**2` will not. That said, I don't have any examples of where the shape would be this complicated. Hence why it is perhaps sufficient to just check if the expression contains a `Variable`

Reply:
I see. The current implementation is the safest option for now, we'll think about something faster the day when this becomes a computational bottleneck.